### PR TITLE
Change `ViewComponent.View()` to flow the `ViewData.Model`.

### PR DESCRIPTION
**Pull request by NTaylorMullen**
- This is more consistent with how controllers work.

#4882

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewComponent.cs [hunk 1]
```diff
@@ -262,7 +262,13 @@ public ViewViewComponentResult View<TModel>(TModel model)
         /// <returns>A <see cref="ViewViewComponentResult"/>.</returns>
         public ViewViewComponentResult View<TModel>(string viewName, TModel model)
         {
-            var viewData = new ViewDataDictionary<TModel>(ViewData, model);
+            // Do not override ViewData.Model unless passed a non-null value.
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewComponent.cs]: This is pokemon comment. It only says what it is.

**Comment by rynowak** [comment]
Are we ok with the breaking change here? This means that model will be non-null in some cases where it was explicitly set to null.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewComponent.cs (hunk 1 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewComponent.cs]: lol, copied from `Controller.cs`. Will nuke in both locations.

**Comment by NTaylorMullen** [comment]
I'd think this would be clarified as a bug since it's inconsistent with Controller.View but i'll let @Eilon decide.

**Comment by rynowak** [comment]
>I'd think this would be clarified as a bug since it's inconsistent with Controller.View but i'll let @Eilon decide.

These are absolutely different cases. 

In `Controller.View(...)` the *current model* is null unless you did something to make it non-null.

In `ViewComponent.View(...)` the *current model* should be the model of the view that called you (if-any).  You could reasonably rely on this behavior in 1.0.0. You might have seen this and infer that this was intended.

I think the proposed fix is a more reasonable behavior, but I also think we'll break some folks.

**Comment by rynowak** [comment]
You also might want to fact check me. If I'm wrong on the view's model doesn't flow to the VC's view then this is all moo :cow2:

**Comment by NTaylorMullen** [comment]
@rynowak Just tried out a sample where in my `ViewComponent` I returned `View()` and in an Index.cshtml I did: `@await Component.Invoke("Foo")`. Index.cshtml had a non-null model but that model did not flow into ViewComponent Foo's View.

**Comment by rynowak** [comment]
OK D O P E

**Comment by Eilon** [comment]
Sounds good to me.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Controller.cs [hunk 2]
```diff
@@ -195,7 +195,6 @@ public virtual PartialViewResult PartialView(object model)
         [NonAction]
         public virtual PartialViewResult PartialView(string viewName, object model)
         {
-            // Do not override ViewData.Model unless passed a non-null value.
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Controller.cs]: This comment is still `true` isn't it?

Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Controller.cs]: Should leave this alone since the same comment remains in `View()` and both are correct.

Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Controller.cs]: https://github.com/aspnet/Mvc/pull/5159#discussion_r75401213

Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Controller.cs]: Then why didn't you nuke the other comment in `View()`?

**Comment by dougbu** [comment]
:shipit: (just one small comment about a comment that others have commented shouldn't be commented)

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Controller.cs (hunk 2 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Controller.cs]: Ah, lookie that. I accidently did the PartialVeiw comment and not the View one. :+1:

**Pull request merged**